### PR TITLE
feat: support Redis connection strings in router

### PR DESCRIPTION
**Pull request by akoenig**
## Motivation and Context

This PR introduces the ability to configure the storage adapter for rate limiting using a Redis connection string. This is an additional feature, ensuring that the existing configuration via a connection tuple remains functional.

```yaml
rate_limit:
  enabled: true
  storage:
    url: "redis://localhost:6379"
```

## TODO

- [ ] Tests or benchmark included
- [ ] Documentation is changed or added on [https://app.gitbook.com/](https://app.gitbook.com/)
- [x] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)

### CODE REVIEW

File: router/core/router.go [hunk 1]
```diff
@@ -1089,10 +1089,24 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
...
+
+		if r.Config.rateLimit.Storage.Url != "" {
+			options, err := redis.ParseURL(r.Config.rateLimit.Storage.Url)
+
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse the redis connection string %s: %w", r.Config.rateLimit.Storage.Url, err)
+			}
+
+			client = redis.NewClient(options)
+		} else {
```
Comment (jensneuse) [inline router/core/router.go:1103]: Will this mean that a combination of URL and password is not possible? Or is a password part of the URL?

**Review by jensneuse — CHANGES_REQUESTED** [review]

**Comment by akoenig** [comment]
Just a thought: Perhaps we could consider deprecating the configuration via `addr` and `password`. The connection string offers more flexibility in the end.

File: router/core/router.go (hunk 1 above)
Comment (akoenig) [reply router/core/router.go:1103]: @jensneuse Password can be a part of the url, e.g. `redis://<user>:<password>@<host>:<port>`

**Comment by jensneuse** [comment]
> Just a thought: Perhaps we could consider deprecating the configuration via `addr` and `password`. The connection string offers more flexibility in the end.

I'm happy with this change.
You can just remove the existing approach.
Also, keep in mind that some tests need an update.

**Comment by akoenig** [comment]
@jensneuse Seems like GitHub Actions issues are solved. Would you mind re-triggering the workflows when you have the chance? Thanks 🙂

File: router/core/router.go [hunk 2]
```diff
@@ -1089,10 +1089,15 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 
 	if r.Config.rateLimit != nil && r.Config.rateLimit.Enabled {
 		handlerOpts.RateLimitConfig = r.Config.rateLimit
-		client := redis.NewClient(&redis.Options{
-			Addr:     r.Config.rateLimit.Storage.Addr,
-			Password: r.Config.rateLimit.Storage.Password,
-		})
+
+		options, err := redis.ParseURL(r.Config.rateLimit.Storage.Url)
+
```
Comment (jensneuse) [inline router/core/router.go:1094]: please remove the whitespace here

File: router/pkg/config/config.schema.json [hunk 3]
```diff
@@ -848,15 +848,10 @@
           "type": "object",
           "additionalProperties": false,
           "properties": {
-            "addr": {
+            "url": {
               "type": "string",
-              "description": "The url of the redis server. The url is specified as a string with the format 'scheme://host:port'.",
-              "format": "url",
```
Comment (jensneuse) [inline router/pkg/config/config.schema.json:854]: Should we keep url here? Or how can we ensure that the input is validated?

**Review by jensneuse — CHANGES_REQUESTED** [review]

**Comment by Aenimus** [comment]
Hi @akoenig,

We'll be addressing this shortly.

Thank you for your patience.

CC: @StarpTech

File: router/pkg/config/config.go [hunk 4]
```diff
@@ -232,8 +232,7 @@ type RateLimitConfiguration struct {
 }
 
 type RedisConfiguration struct {
-	Addr      string `yaml:"addr" default:"localhost:6379" envconfig:"REDIS_ADDR"`
-	Password  string `yaml:"password,omitempty" envconfig:"REDIS_PASSWORD"`
+	Url       string `yaml:"url,omitempty" default:"redis://localhost:6379" envconfig:"REDIS_URL"`
```
Comment (StarpTech) [inline router/pkg/config/config.go:235]: I propose to prefix all environment variables with `RATE_LIMIT_` to make the position clear in the configuration.

```
RATE_LIMIT_REDIS_URL
RATE_LIMIT_REDIS_KEY_PREFIX
```

File: router/core/router.go [hunk 5]
```diff
@@ -1089,10 +1089,13 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 
 	if r.Config.rateLimit != nil && r.Config.rateLimit.Enabled {
 		handlerOpts.RateLimitConfig = r.Config.rateLimit
-		client := redis.NewClient(&redis.Options{
-			Addr:     r.Config.rateLimit.Storage.Addr,
-			Password: r.Config.rateLimit.Storage.Password,
-		})
+		options, err := redis.ParseURL(r.Config.rateLimit.Storage.Url)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse the redis connection string %s: %w", r.Config.rateLimit.Storage.Url, err)
```
Comment (StarpTech) [inline router/core/router.go:1094]: Because the string can contain sensitive information, we should not log it. The error message will be logged.

**Review by StarpTech — CHANGES_REQUESTED** [review]

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**